### PR TITLE
GFORMS-1408 - Reusable expressions

### PR DESCRIPTION
**Pull request by josef-vlach**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala [hunk 1]
```diff
@@ -60,11 +60,13 @@ class FormTemplateService(formTemplateRepo: Repo[FormTemplate], formTemplateRawR
         }
       })
 
-  def verifyAndSave(formTemplate: FormTemplate): FOpt[Unit] =
+  def verifyAndSave(formTemplate: FormTemplate)(expressionsContext: Substitutions): FOpt[Unit] = {
+    val substitutedFormTemplate = substituteExpressions(formTemplate, expressionsContext)
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala:64]: If you import ```import uk.gov.hmrc.gform.formtemplate.Substituter.SubstituterSyntax```

this line can be

```
val substitutedFormTemplate = formTemplate(expressionsContext)
```

We may not need that trait `trait SubstituteExpressions`

File: app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          case JsObject(fields) =>
+            fields.toList.traverse { case (k, v) =>
+              val expressionId = ExpressionId(k)
+              val maybeExpr: Opt[Expr] = v match {
+                case JsString(exprStr) =>
+                  val parsed: Opt[ValueExpr] = ValueParser.validate("${" + exprStr + "}")
+
+                  parsed match {
+                    case Right(TextExpression(expr)) => Right(expr)
+                    case Right(_)                    => Left(UnexpectedState("Wrong expression: " + v))
```
[the remark this answers is not in the document]
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala:45]: Good question, this line will never be hit, as `parsed` is guaranteed to be `TextExpression` due to `${..}` on line 41 .
If this line is ever hit, there is something wrong somewhere else.

Also there is no expression to substitute in `DateExpression` or `ChoiceExpression`

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala (hunk 1 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala:64]: I don't want to make `SubstituterSyntax` widely available so I rather make it `private`

File: app/uk/gov/hmrc/gform/formtemplate/Substituter.scala [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+      case Period(l, r)      => Period(substitute(substitutions, l), substitute(substitutions, r))
+      case Sum(l)            => Sum(substitute(substitutions, l))
+      case PeriodExt(p, pe)  => PeriodExt(substitute(substitutions, p), pe)
+      case DateCtx(dateExpr) => DateCtx(dateExpr(substitutions))
+      case i @ IfElse(cond, l, r) =>
+        IfElse(cond(substitutions), substitute(substitutions, l), substitute(substitutions, r))
+      case f @ FormCtx(formComponentId) =>
+        // Replace FormComponentId with top level expression if one exists
+        substitutions.expressions.getOrElse(ExpressionId(formComponentId.value), f)
+      case otherwise => otherwise
```
[the remark this answers is not in the document]
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/Substituter.scala:71]: It does, good suggestion.

File: app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            Left(
+              UnexpectedState(
+                "Field 'expressions' needs to be a JsObject, but got " + nonJsObject.getClass.getSimpleName + ": " + nonJsObject
+              )
+            )
+        }
+      maybeSubstitutions.map(_.toMap).map(Substitutions.apply)
+    }
+
+    templateRaw.value \ "expressions" match {
```
[the remark this answers is not in the document]
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala:69]: I think explicit pattern matching is better for readability in this case.

File: app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/formtemplate/Substitutions.scala:45]: Thanks for the detailed explanation

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala:64]: Ok. It did cause a bit of confusion when I saw the usage for the first time in Substituter class. The IDE was not helping either

**Review by girishkamat — APPROVED** [review]

**Pull request merged**